flow/project.py

```diff
@@ -291,53 +316,43 @@ def __str__(self):
 
     def __repr__(self):
         return "{type}(name='{name}', job='{job}', cmd={cmd}, directives={directives})".format(
-            type=type(self).__name__,
-            name=self.name,
-            job=str(self.job),
-            cmd=repr(self.cmd),
-            directives=self.directives)
+                   type=type(self).__name__,
+                   name=self.name,
+                   job=str(self.job),
+                   cmd=repr(self.cmd),
+                   directives=self.directives)
 
-    def get_id(self, index=0):
-        "Return a name, which identifies this job-operation."
-        project = self.job._project
-
-        # The full name is designed to be truly unique for each job-operation.
-        full_name = '{}%{}%{}%{}'.format(
-            project.root_directory(), self.job.get_id(), self.name, index)
-
-        # The job_op_id is a hash computed from the unique full name.
-        job_op_id = calc_id(full_name)
-
-        # The actual job id is then constructed from a readable part and the job_op_id,
-        # ensuring that the job-op is still somewhat identifiable, but guaranteed to
-        # be unique. The readable name is based on the project id, job id, operation name,
-        # and the index number. All names and the id itself are restricted in length
-        # to guarantee that the id does not get too long.
-        max_len = self.MAX_LEN_ID - len(job_op_id)
-        if max_len < len(job_op_id):
-            raise ValueError("Value for MAX_LEN_ID is too small ({}).".format(self.MAX_LEN_ID))
+    def __hash__(self):
+        return int(sha1(self.id.encode('utf-8')).hexdigest(), 16)
 
-        readable_name = '{}/{}/{}/{:04d}/'.format(
-            str(project)[:12], str(self.job)[:8], self.name[:12], index)[:max_len]
+    def __eq__(self, other):
+        return self.id == other.id
 
-        # By appending the unique job_op_id, we ensure that each id is truly unique.
-        return readable_name + job_op_id
+    def get_id(self):
+        return self._id
 
-    def __hash__(self):
-        return int(sha1(self.get_id().encode('utf-8')).hexdigest(), 16)
+    @property
+    def id(self):
+        return self._id
 
-    def __eq__(self, other):
-        return self.get_id() == other.get_id()
+    @property
+    def cmd(self):
+        if callable(self._cmd):
+            return self._cmd()
```

Comment:
This prevents errors when an operation that doesn't fork is run without a valid command. This could happen if the entrypoint is underspecified which can occur easily in our Python API.

Reply:
I honestly don't follow that and it smells like a design flaw to me, but I'll trust your judgement.

Reply:
The other solution(s) are to try and guess an entrypoint at instantiation (but then if that is different from the script file where main is called we will error) or error if an entrypoint is not provided at instantiation (which would break backwards compatibility).

Reply:
I think it would be very important to add an in-code comment here to clarify this, because other people who read this will also be confused about that.